File: tetrattack_generator/js/t_model.js

const colors = ['red', 'yellow', 'blue', 'green', 'pink', 'black', 'orange'];

/* board:
 * {
 *  num_rows: 0,
 *  num_columns: 0,
 *  cells: Map
 * }
 */

/* pos: { x: 3, y: 9 } */

function prepare_swap(pos1, pos2) {
    return function swap(board) {
        color1 = board.cells.get(pos1);
        color2 = board.cells.get(pos2);
        board.cells.put(pos2, color1);
        board.cells.put(pos1, color2);
        return board;
    };
}

function prepare_remove(pos) {
    return function remove(board) {
        board.cells.set(pos, null);
        return board;
    };
}

function prepare_move(pos_source, pos_destination) {
    return function move(board) {
        const color = board.cells.get(pos_source);
        const board1 = prepare_remove(pos_source)(board);
        return prepare_insert(pos_destination, color)(board);
    };
}

function prepare_insert(pos, color) {
    return function insert(board) {
        if (board.cells.get(dest_pos)) {
            prepare_move(pos, pos_above(pos))(board);
        }
        board.cells.set(pos, color);
        return board;
    };
}

function pos_add(pos1, pos2) {
    if (!pos1) {
        return pos2;
    }
    if (!pos2) {
        return pos1;
    }
    return {
        x: pos1.x + pos2.x,
        y: pos1.y + pos2.y
    };
}

function pos_above(pos) {
    return pos_add(pos, {x: 0, y: 1});
}

function pos_below(pos) {
    return pos_add(pos, {x: 0, y: -1});
}

function pos_left(pos) {
    return pos_add(pos, {x: -1, y: 0});
}

function pos_below(pos) {
    return pos_add(pos, {x: 1, y: 0});
}

File: tetrattack_generator/t_model.hs

module TetrattackBoard
( Action
, Board(..)
, Color(..)
, Pos
, board_cell
, board_insert
, board_move_and_insert
, board_swap
, no_action
) where

import Data.Map (Map)
import qualified Data.Map as Map

{-
( Puzzle (..)
, PuzzleCell
, ProtoCell
, puzzleToBoard
, puzzleUnwrapDirs
) where
-}

data Color = Red | Yellow | Blue | Green | Pink | Black | Orange
data Board = Board {
    board_num_rows :: Int,
    board_num_columns :: Int,
    board_cells :: Map Pos Color
}
type Action = Board -> Board
type Pos = (Int, Int)
data Direction = DUp | DDown | DLeft | DRight

no_action = (\board -> board)

map_unconditional_assign :: Ord k => k -> Maybe v -> Map k v -> Map k v
map_unconditional_assign key Nothing mappy = Map.delete key mappy
map_unconditional_assign key (Just value) mappy = Map.insert key value mappy

pos_direction :: Direction -> Pos -> Pos
pos_direction DUp (x, y) = (x, y + 1)
pos_direction DLeft (x, y) = (x - 1, y)
pos_direction DRight (x, y) = (x + 1, y)
pos_direction DDown (x, y) = (x, y - 1)

board_swap :: Pos -> Pos -> Board -> Board
board_swap pos1 pos2 board = board_set_cell color2 pos1 (board_set_cell color1 pos2 board)
    where color1 = board_cell pos1 board
          color2 = board_cell pos2 board

board_move_and_insert :: Pos -> Pos -> Board -> Board
board_move_and_insert source_pos dest_pos board = board_insert source_cell dest_pos (board_delete source_pos board)
    where source_cell = board_cell source_pos board

board_insert :: Maybe Color -> Pos -> Board -> Board
board_insert color pos board = board_set_cell color pos shifted_board
    where pos_above = pos_direction DUp pos
          color_above = board_cell pos_above board
          shifted_board = case color_above of
                            Nothing -> board
                            Just c -> board_insert color_above pos_above board

board_delete :: Pos -> Board -> Board
board_delete pos board = board { board_cells = Map.delete pos (board_cells board) }

board_cell :: Pos -> Board -> Maybe Color
board_cell pos board = Map.lookup pos (board_cells board)

board_set_cell :: Maybe Color -> Pos -> Board -> Board
board_set_cell color pos board = board { board_cells = map_unconditional_assign pos color (board_cells board) }
